### ISSUE TITLE
Printed representation of Coalton types is not readable

### ISSUE THREAD

**Issue by gefjon** [coalton-lang/coalton#707]
The printed representation of Coalton types uses the `#.` macro in an attempt to be readable, but does not quote subobjects. This works fine for subobjects which are also Coalton types, or self-evaluating CL types, but breaks with lists (and would probably break with `repr :native symbol` objects).

E.g.

```
(coalton-toplevel
  (define-type ContainsList
    (ContainsList (List UFix))))

(cl:prin1-to-string (coalton (ContainsList (make-list 0 1 2))))
; => "#.(CONTAINSLIST (0 1 2))"

(cl:read-from-string (cl:prin1-to-string (coalton (ContainsList (make-list 0 1 2)))))
; in: CONTAINSLIST (0 1 2)
;     (0 1 2)
; 
; caught COMMON-LISP:ERROR:
;   illegal function call
; 
; compilation unit finished
;   caught 1 ERROR condition
```

Easy solution: stick a quote in front of each argument, so that `ContainsList` prints as `#.(CONTAINSLIST '(0 1 2))`.

**Comment by stylewarning** [comment]
I think the whole printing thing needs to be changed. I've never copied a printed coalton value to read. It also seems faulty to assume everything is readable by default.

**Comment by gefjon** [comment]
> I've never copied a printed coalton value to read.

I tried for the first time today; I had some slow experiments which I queued up and ran over the weekend, with each one dumping its output (as a `define-type` struct) to a file, so that I could come back later, read them in and process them. I ended up manually going in to each dumped text file and quoting all the lists, and after doing that I was able to re-construct the results with `cl:read` (well, actually `uiop:read-file-form`, but whatever).

I think automatically being able to serialize and deserialize data to and from text is super useful; I wouldn't want to lose that ability compared to Common Lisp. In the long term, I think it would be real slick to have a typeclass `Serialize` with instances derived by the compiler. (And not derived for certain stateful/resource-ey types, e.g. `Cell`, and maybe `HashTable`, or types that contain them.)

In the short term, though, our `print-object` methods are clearly trying to print in a readable format, otherwise they wouldn't be `#.`'ed. Reading printed objects has correctness pitfalls, but the only way to do it is by going through Common Lisp, which has pitfalls of its own, so I'm not super concerned. I'd like to make it so our printed representations are actually readable for now, and worry about correctness in the future, once we get support for things like file i/o without shelling out to CL.
